scipy/sparse/linalg/_isolve/iterative.py

```diff
@@ -1,496 +1,636 @@
-        # emulate old legacy behavior
-        resid = get_residual()
-        if resid <= tol:
-            return 'exit'
-        if bnrm2 == 0:
-            return tol
-        else:
-            return tol * float(bnrm2)
-    else:
-        return max(float(atol), tol * float(bnrm2))
-
-
-def set_docstring(header, Ainfo, footer='', atol_default='0'):
-    def combine(fn):
-        fn.__doc__ = '\n'.join((header, common_doc1,
-                                '    ' + Ainfo.replace('\n', '\n    '),
-                                common_doc2, dedent(footer)))
-        return fn
-    return combine
-
-
-@set_docstring('Use BIConjugate Gradient iteration to solve ``Ax = b``.',
-               'The real or complex N-by-N matrix of the linear system.\n'
-               'Alternatively, ``A`` can be a linear operator which can\n'
-               'produce ``Ax`` and ``A^T x`` using, e.g.,\n'
-               '``scipy.sparse.linalg.LinearOperator``.',
-               footer="""\
-               Examples
-               --------
-               >>> import numpy as np
-               >>> from scipy.sparse import csc_matrix
-               >>> from scipy.sparse.linalg import bicg
-               >>> A = csc_matrix([[3, 2, 0], [1, -1, 0], [0, 5, 1]], dtype=float)
-               >>> b = np.array([2, 4, -1], dtype=float)
-               >>> x, exitCode = bicg(A, b)
-               >>> print(exitCode)            # 0 indicates successful convergence
-               0
-               >>> np.allclose(A.dot(x), b)
-               True
-
-               """
-               )
-@non_reentrant()
-def bicg(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None, atol=None):
-    A,M,x,b,postprocess = make_system(A, M, x0, b)
+    atol = max(float(atol), float(rtol) * float(np.linalg.norm(b)))
 
     n = len(b)
+    if (np.iscomplexobj(A) or np.iscomplexobj(b)):
+        dotprod = np.vdot
+    else:
+        dotprod = np.dot
+
     if maxiter is None:
         maxiter = n*10
 
     matvec, rmatvec = A.matvec, A.rmatvec
     psolve, rpsolve = M.matvec, M.rmatvec
-    ltr = _type_conv[x.dtype.char]
-    revcom = getattr(_iterative, ltr + 'bicgrevcom')
-
-    def get_residual():
-        return np.linalg.norm(matvec(x) - b)
-    atol = _get_atol(tol, atol, np.linalg.norm(b), get_residual, 'bicg')
-    if atol == 'exit':
-        return postprocess(x), 0
-
-    resid = atol
-    ndx1 = 1
-    ndx2 = -1
-    # Use _aligned_zeros to work around a f2py bug in Numpy 1.9.1
-    work = _aligned_zeros(6*n,dtype=x.dtype)
-    ijob = 1
-    info = 0
-    ftflag = True
-    iter_ = maxiter
-    while True:
-        olditer = iter_
-        x, iter_, resid, info, ndx1, ndx2, sclr1, sclr2, ijob = \
-           revcom(b, x, work, iter_, resid, info, ndx1, ndx2, ijob)
-        if callback is not None and iter_ > olditer:
+
+    rhotol = np.finfo(x.dtype.char).eps**2
+
+    # Is there any tolerance set? since b can be all 0.
+    if atol == 0.:
+        atol = 1e-5
```

Comment:
Is this not redundant given the `atol = max(float(atol), float(rtol) * float(np.linalg.norm(b)))` earlier? Same in the other functions

Reply:
That's right, I guess this is what happens when you iterate 15 different versions of tolerance handling :)